### PR TITLE
chore: catboost, detectron, sklearn partial kwargs

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/catboost.py [hunk 1]
```diff
@@ -244,24 +251,27 @@ def num_concurrency_per_replica(self) -> int:
...
         self._predict_fn = getattr(self._model, self._predict_fn_name)
 
-    # pylint: disable=arguments-differ
-    def _run_batch(  # type: ignore[override]
-        self, input_data: t.Union[np.ndarray, "pd.DataFrame", cbt.Pool]
-    ) -> np.ndarray:
-        # Take a batch type
-        # Take a containers
-        res = self._predict_fn(input_data)
+    def _run_batch(
```
Comment (parano) [inline bentoml/catboost.py:259]: catboost's `predcit` method only accept one argument as input, there's no need for adding *args and **kwargs in this case

Comment (parano) [reply bentoml/catboost.py:259]: see https://github.com/catboost/catboost/blob/master/catboost/python-package/catboost/core.py#L4767

File: bentoml/detectron.py [hunk 2]
```diff
@@ -208,33 +218,42 @@ def num_replica(self) -> int:
...
-        self, input_data: t.Union[np.ndarray, torch.Tensor]
-    ) -> np.ndarray:
-        images = np.split(input_data, input_data.shape[0], 0)
-        images = [image.squeeze(axis=0) for image in images]
-        if isinstance(input_data, np.ndarray):
-            images = [torch.from_numpy(image) for image in images]
+    def _run_batch(
+        self,
+        *args: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", torch.Tensor],
+        **kwargs: t.Any,
```
Comment (parano) [inline bentoml/detectron.py:233]: does detectron model's `__call__` accept kwargs? if not, we don't need to add this

File: bentoml/sklearn.py [hunk 3]
```diff
@@ -130,38 +140,37 @@ def __init__(
...
-    def _run_batch(  # type: ignore[override]
+    def _run_batch(
         self,
-        *args: t.Union[np.ndarray, "pd.DataFrame"],
-        **kwargs: t.Union[np.ndarray, "pd.DataFrame"],
-    ) -> "np.ndarray":
-        func = getattr(self._model, self._function_name)
-        with self._parallel_ctx:
-            return func(*args, **kwargs)
+        *args: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", "DataFrame"],
```
Comment (parano) [inline bentoml/sklearn.py:166]: I think it's the same case for sklearn, this should be just:
```
def _run_batch(self, batch_input: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", "DataFrame"])
```

Comment (aarnphm) [reply bentoml/sklearn.py:166]: noted

File: bentoml/detectron.py (hunk 2 above)
Comment (aarnphm) [reply bentoml/detectron.py:233]: we actually want to support kwargs in this case. https://github.com/facebookresearch/detectron2/blob/main/projects/Panoptic-DeepLab/panoptic_deeplab/panoptic_seg.py

File: bentoml/catboost.py [hunk 4]
```diff
@@ -244,24 +250,24 @@ def num_concurrency_per_replica(self) -> int:
...
-        self, input_data: t.Union[np.ndarray, "pd.DataFrame", cbt.Pool]
-    ) -> np.ndarray:
-        # Take a batch type
-        # Take a containers
-        res = self._predict_fn(input_data)
+    def _run_batch(  # type: ignore[reportIncompatibleMethodOverride]
+        self,
+        inputs: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", "DataFrame", cbt.Pool],
+    ) -> "np.ndarray[t.Any, np.dtype[t.Any]]":
+        # TODO: Take a batch type and containers
```
Comment (parano) [inline bentoml/catboost.py:263]:
```suggestion
```

File: bentoml/sklearn.py [hunk 5]
```diff
@@ -130,38 +139,33 @@ def __init__(
...
-            "threading", n_jobs=self.num_concurrency_per_replica
-        )
+        self._backend = "loky"
         self._function_name = function_name
 
     @property
     def num_concurrency_per_replica(self) -> int:
-        # NOTE: sklearn doesn't use GPU, so return max. no. of CPU's.
-        return int(round(self.resource_quota.cpu))
+        return 1
```
Comment (parano) [inline bentoml/sklearn.py:147]: revert num_concurrency_per_replica and  num_replica

File: bentoml/sklearn.py [hunk 6]
```diff
@@ -49,7 +59,7 @@ def _get_model_info(
 def load(
     tag: t.Union[str, Tag],
     model_store: "ModelStore" = Provide[BentoMLContainer.model_store],
-) -> _MT:
+) -> t.Any:
```
Comment (parano) [inline bentoml/sklearn.py:62]: look into sklearn see if there's a base model class

File: bentoml/detectron.py (hunk 2 above)
Comment (parano) [reply bentoml/detectron.py:234]:
```suggestion
        **kwargs: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", torch.Tensor],
```

File: bentoml/detectron.py [hunk 7]
```diff
@@ -183,7 +194,7 @@ class _DetectronRunner(Runner):
     @inject
     def __init__(
         self,
-        tag: str,
+        tag: t.Union[str, Tag],
```
Comment (parano) [inline bentoml/detectron.py:197]: todo: add partial args

File: bentoml/catboost.py [hunk 8]
```diff
@@ -244,24 +250,23 @@ def num_concurrency_per_replica(self) -> int:
...
     # pylint: disable=arguments-differ
-    def _run_batch(  # type: ignore[override]
-        self, input_data: t.Union[np.ndarray, "pd.DataFrame", cbt.Pool]
-    ) -> np.ndarray:
-        # Take a batch type
-        # Take a containers
-        res = self._predict_fn(input_data)
+    def _run_batch(  # type: ignore[reportIncompatibleMethodOverride]
+        self,
+        inputs: t.Union["np.ndarray[t.Any, np.dtype[t.Any]]", "DataFrame", cbt.Pool],
```
Comment (parano) [inline bentoml/catboost.py:261]: todo: cbt.Pool data container

File: bentoml/detectron.py [hunk 9]
```diff
@@ -208,33 +219,42 @@ def num_replica(self) -> int:
...
+                return torch.from_numpy(item)
+            return item
+
+        params = params.map(_mapping)
 
-        inputs = [dict(image=image) for image in images]
+        inputs = [{"image": image} for image in params.args]
 
-        res = self._predict_fn(inputs)
+        res = self._predict_fn(inputs, **params.kwargs)
```
Comment (parano) [inline bentoml/detectron.py:251]: change this to partial kwargs

File: bentoml/sklearn.py (hunk 6 above)
Comment (larme) [reply bentoml/sklearn.py]: let's try with t.Union[BaseEstimator, Pipeline]

File: bentoml/detectron.py (hunk 7 above)
Comment (larme) [reply bentoml/detectron.py:205]: I'll do this later

**Review by larme — APPROVED** [review]

**Review by larme — APPROVED** [review]

**Pull request merged**